mlkem/poly.c

```diff
@@ -108,13 +109,12 @@ void poly_decompress(poly *r, const uint8_t a[KYBER_POLYCOMPRESSEDBYTES])
 {
     #if (KYBER_POLYCOMPRESSEDBYTES == 128)
     for (int i = 0; i < KYBER_N / 2; i++)
-        __CPROVER_assigns(i, __CPROVER_object_whole(r))
-        __CPROVER_loop_invariant(i >= 0)
-        __CPROVER_loop_invariant(i <= KYBER_N / 2)
-        __CPROVER_loop_invariant(__CPROVER_forall { unsigned int k; (k < i) ==>
-                                 ( r->coeffs[2 * k]     >= 0 && r->coeffs[2 * k]     < KYBER_Q &&
-                                   r->coeffs[2 * k + 1] >= 0 && r->coeffs[2 * k + 1] < KYBER_Q ) })
-        __CPROVER_decreases(KYBER_N / 2 - i)
+        ASSIGNS(i, OBJECT_WHOLE(r))
+        INVARIANT(i >= 0)
+        INVARIANT(i <= KYBER_N / 2)
+        INVARIANT(FORALL(int, k, 0, (i-1), (r->coeffs[2 * k]     >= 0 && r->coeffs[2 * k]     < KYBER_Q &&
```

Comment:
Can this not be formulated using `ARRAY_IN_TYPE`?

Reply:
I don't think so... the quantified variable is "k" but we want "2 * k" and so on inside the predicate on the right, so it wouldn't expand correctly using the macro.

Reply:
You'd need to double the upper bound index, yes, but wouldn't that work?

Reply:
Don't think so... if the quantified variable is "k", then the macro produces r->coeffs[k] on the right, not r->coeffs[2 * k]

Reply:
But you have two clauses, one for `2k` and one for `2k+1` -- so overall you bound _all_ array elements..

Reply:
OK... got it and it works.